integ/lib/testing-tier.ts

```diff
@@ -0,0 +1,340 @@
+import { CfnOutput, Construct, Duration, Stack, StackProps } from '@aws-cdk/core';
+import { X509CertificatePem, MongoDbInstaller, MongoDbSsplLicenseAcceptance, MongoDbVersion } from 'aws-rfdk';
+import { IWorkerFleet, RenderQueue, Repository } from 'aws-rfdk/deadline';
+import { RenderStruct } from './render-struct';
+import { StorageStruct, IRenderFarmDb } from './storage-struct';
+import { WorkerStruct } from './worker-struct';
+
+// Interface for choosing userData installation options
+interface UserDataConfigProps {
+  installDeadlineClient: boolean;
+  fetchDocdbCert: boolean;
+  testingScriptPath: string;
+}
+
+// Params object for TestingTier
+export interface TestingTierProps extends StackProps {
+  integStackTag: string;
+}
+
+// Class constructor
+export class TestingTier extends Stack {
+
+  private testInstance: BastionHostLinux;
+  private deadlineVersion: string = process.env.DEADLINE_VERSION!.toString();
+  private stagePath: string = process.env.DEADLINE_STAGING_PATH!.toString();
+
+  constructor(scope: Construct, id: string, props: TestingTierProps) {
+    super(scope, id, props);
+
+    // Collect environment variables
+    const infrastructureStackName = 'RFDKIntegInfrastructure' + props.integStackTag;
+
+    // Vpc.fromLookup acquires vpc deployed to the _infrastructure stack
+    const vpc = Vpc.fromLookup(this, 'Vpc', { tags: { StackName: infrastructureStackName }}) as Vpc;
+
+    // Create an instance that can be used for testing; SSM commands are communicated to the
+    // host instance to run test scripts installed during setup of the instance
+    const testInstance: BastionHostLinux = new BastionHostLinux(this, 'Bastion', {
+      vpc,
+      instanceType: new InstanceType('t3.small'),
+      subnetSelection: {
+        subnetType: SubnetType.PRIVATE,
+      },
+    });
+    this.testInstance = testInstance;
+
+    // Output bastion id for use in tests
+    new CfnOutput(this, 'bastionId', {
+      value: testInstance.instanceId,
+    });
+
+  }
+
+  // Each test suite has an associated public function which calls on several of this class's private functions to set up the test bastion
+  public configureRepositoryTest(structs:Array<StorageStruct>) {
+    structs.forEach( storageStruct => {
+
+      const testSuiteId = 'DL' + (structs.indexOf(storageStruct) + 1).toString();
+
+      const repo = storageStruct.repo;
+      this.configureRepo(testSuiteId, repo);
+
+      const database = storageStruct.database;
+      this.configureDatabase(testSuiteId, database);
+
+      const efs = storageStruct.efs;
+      this.configureEfs(efs);
+
+      const cert = storageStruct.database.cert;
+      this.configureCert(testSuiteId, cert);
+    });
+
+    this.configureBastionUserData({
+      installDeadlineClient: false,
+      fetchDocdbCert: true,
+      testingScriptPath: '../components/deadline/deadline_01_repository/scripts/bastion/testing',
+    });
+
+    this.configureMongodb();
+  }
```

Comment:
It appears that this method and the two below it are test-specific. Instead of putting the logic in this common Stack, what if each test sub-classed this stack? This way we get to re-use the common code, but separate the test-specific code. The logic in this method could be moved into the constructor of the repository test's sub-class. To illustrate:

```ts
import {
  TestingTier,
  TestingTierProps,
} from '../../../../lib/testing-tier'

export interface RepositoryTestingTierProps extends TestingTierProps {
  readonly structs: Array<StorageStruct>;
}

export class RepositoryTestingTier extends TestingTeir {
  constructor(scope: Construct, id: string, props: RepositoryTestingTierProps) {
    super(scope, id, props);

    // Move configureRepositoryTest code into here
  }
}
```

Reply:
I think that makes a lot of sens, I'll take a look at revising this today.

Reply:
Much better! Resolved.